### PR TITLE
Feature/update draw sonar

### PR DESCRIPTION
**Pull request by amarburg**
Major rewrite of internals to improve performance and image quality.

I have tried to maintain backwards compatibility both in terms of data output e.g. image size as a function of # of bearings/bins and API, but have not been too precious.  Please flag if output is not as expected.

Currently an "early access" draft, still making improvements.

### CODE REVIEW

**Comment by amarburg** [comment]
OK, we're now at "alpha release" stage.  Still work to do, but clean-ish.    As far as I know, the ROS interfaces are unchanged.  The C++ API for drawSonar is also the same ... 

I **have** changed the virtual requirements for AbstractSonarInterface -- it's internally consistent within this repo but may break some stuff elsewhere.

File: README.md [hunk 1]
```diff
@@ -2,7 +2,83 @@
...
+
+The topic `drawn_sonar` (a [sensor_msgs/Image](https://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/Image.html)) contains an image of the sonar drawn with a cartesian projection
+with the origin (range = 0) of the sonar centered on the bottom edge of
+the image, and azimuth=0 extending vertically upwards in the image.  By default,
+the image height is set by the number range bins, and the image width is
+automatically determined based on the height and the min/max azimuth of the
+sonar image.   The color map used to convert the sonar intensity to RGB is set
+in code.
+
+![](drawn_sonar.png)
```
Comment (lindzey) [inline README.md:33]: It looks like the image you chose for this was from the *old* algorithm?

**Comment by lindzey** [comment]
@amarburg -- this is a rather large PR and I'm having trouble figuring out where to start. 

It would be really helpful (and good practice in general) if you could include a narrative description of the changes you made and why as a comment to the PR.

**Comment by amarburg** [comment]
> @amarburg -- this is a rather large PR and I'm having trouble figuring out where to start.
> 
> It would be really helpful (and good practice in general) if you could include a narrative description of the changes you made and why as a comment to the PR.

Sure, it did get a bit far-ranging.

Major changes as follows:

* Broke the drawSonar functionality into two parts:   `drawSonarRect` applies the color map to render the sonar data as a 2D (numAzimuths x numRanges) image.   `drawSonar` then users `cv::remap` to warp that 2D image into the "fan" shape.  This is the major update to the aesthetics and performance of the code.
* Moved both of above from a function-based interface to class members in `SonarDrawer` ... this allows persisting of maps between drawing operations when the sonar geometry doesn't change.   Left the functional API in place, it's now a trivial wrapper which creates an SonarDrawer, calls it, then destroys it.
* Changed AbstractSonarInterface s.t. its primary interface requirement for inherited classes is functions which return const std::vector<float> & of azimuths and range bins.  Added functions for searching for and caching minimum and maximum azimuth/range values from these vectors.
* Rewrote SonarColorMap with templated lookup functions to return appropriately scaled pixel values for either Vec3f (in range [0,1]) or Vec3b (in range [0,255]).  This cleans up the drawSonar code and removes arbitrary "255 *"s

Minor / induced changes:

* Restructuing of the sonar_draw_nodelet
* Added AbstractSonarInterface.cpp partly to hold a new static const member, and also to include a few slightly more complex member functions.
* Moved the old functions into their own source file, an a new namespace sonar_image_proc::old_api
* Moved SonarImageMsgInterface into a new ROS-only header file ins src_ros to reduce the size of sonar_draw_nodelet

File: README.md [hunk 2]
```diff
@@ -2,7 +2,83 @@
...
+
+![](drawn_sonar_rect.png)
+
+If the param `publish_timing` is `true`, the node will track the elapsed time to
+draw each sonar image and publish that information to `sonar_image_proc_timing`
+as a
+[std_msgs/String](http://docs.ros.org/en/noetic/api/std_msgs/html/msg/String.html)
+containing a JSON dict.
+
+If the param `publish_old` is `true`, the node will also draw the sonar using
```
[the remark this answers is not in the document]
Comment (amarburg) [reply README.md:57]: Partly to help with review.  I wanted to ensure I didn't introduce e.g. a left-right mirror in the drawing function.

File: README.md (hunk 1 above)
Comment (amarburg) [reply README.md:33]: doh

File: README.md [hunk 3]
```diff
@@ -2,7 +2,83 @@
...
+## Publishers
+
+By default publishes two topics
+
+The topic `drawn_sonar` (a [sensor_msgs/Image](https://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/Image.html)) contains an image of the sonar drawn with a cartesian projection
+with the origin (range = 0) of the sonar centered on the bottom edge of
+the image, and azimuth=0 extending vertically upwards in the image.  By default,
+the image height is set by the number range bins, and the image width is
+automatically determined based on the height and the min/max azimuth of the
+sonar image.   The color map used to convert the sonar intensity to RGB is set
```
[the remark this answers is not in the document]
Comment (amarburg) [reply README.md:30]: Yeah.   Also proper handling for > 8 bit data.

[the remark this answers is not in the document]
Comment (amarburg) [reply README.md:30]: Pushed these two as issues...

File: README.md (hunk 1 above)
Comment (amarburg) [reply README.md:33]: Corrected this ... still not the most exciting image.   May replace with an image of EFL when I have the chance...

**Comment by lindzey** [comment]
Based on our discussion this morning, I'm merging this!

* We checked that it was consistent with the old version's output (just faster and better resolution)
* We checked that it worked with new bagfiles from the tank
* We double-checked that azimuth signs were correct, and that apparent sonar motion matched camera data

**Comment by lindzey** [comment]
Err... I'll be merging it as soon as I sign in to my other github account...

**Pull request merged**